<u>**SMA Project Check list**</u>: <br>
https://www.notion.so/mekongdelta/SMA-project-checklist-1bd9d61bcb894b7eaa7e3958d6a1bbae

# Problem Statement

TO BE ADDED AFTERWARDS

# Data Sources

In [1]:
# install main library YFinance
!pip install yfinance

In [16]:
# IMPORTS
import numpy as np
import pandas as pd

#Fin Data Sources
import yfinance as yf
import pandas_datareader as pdr

#Data viz
import plotly.graph_objs as go
import plotly.express as px

import time
from datetime import date


## Selecting the period for the observation

In [3]:
# Set end to the last date of March 2024
end = date(year=2024, month=3, day=31)
print(f'Year = {end.year}; month= {end.month}; day={end.day}')

# Set start to April 1, 2000
start = date(year=2000, month=4, day=1)
print(f'Period for indexes: {start} to {end} ')

Year = 2024; month= 3; day=31
Period for indexes: 2000-04-01 to 2024-03-31 


In [4]:
nifty_50_df = yf.download(tickers = "^NSEI",
            start=start,
            end = end,
            interval = "1d")

[*********************100%%**********************]  1 of 1 completed


In [5]:
nifty_50_df.head()

Open         High          Low        Close    Adj Close  \
Date                                                                          
2007-09-17  4518.450195  4549.049805  4482.850098  4494.649902  4494.649902   
2007-09-18  4494.100098  4551.799805  4481.549805  4546.200195  4546.200195   
2007-09-19  4550.250000  4739.000000  4550.250000  4732.350098  4732.350098   
2007-09-20  4734.850098  4760.850098  4721.149902  4747.549805  4747.549805   
2007-09-21  4752.950195  4855.700195  4733.700195  4837.549805  4837.549805   

            Volume  
Date                
2007-09-17       0  
2007-09-18       0  
2007-09-19       0  
2007-09-20       0  
2007-09-21       0

In [6]:
nifty_50_df.tail()

Open          High           Low         Close  \
Date                                                                 
2024-03-21  21989.900391  22080.949219  21941.300781  22011.949219   
2024-03-22  21932.199219  22180.699219  21883.300781  22096.750000   
2024-03-26  21947.900391  22073.199219  21947.550781  22004.699219   
2024-03-27  22053.949219  22193.599609  22052.849609  22123.650391   
2024-03-28  22163.599609  22516.000000  22163.599609  22326.900391   

               Adj Close  Volume  
Date                              
2024-03-21  22011.949219  353200  
2024-03-22  22096.750000  388700  
2024-03-26  22004.699219  328400  
2024-03-27  22123.650391  409100  
2024-03-28  22326.900391  410500

`NOTE`: The date for which the yfincae has the data for nifty 50 is from 17,9,2007. So we are changing the period of observation from `April 1, 2024` to `April 1,2009` which is just after the 2008 financial crisis

In [7]:
# Set end to the last date of March 2024
end = date(year=2024, month=3, day=31)
print(f'Year = {end.year}; month= {end.month}; day={end.day}')

# Set start to April 1, 2009
start = date(year=2009, month=4, day=1)
print(f'Period for indexes: {start} to {end} ')

Year = 2024; month= 3; day=31
Period for indexes: 2009-04-01 to 2024-03-31 


In [8]:
nifty_50_df = yf.download(tickers = "^NSEI",
            start=start,
            end = end,
            interval = "1d")

[*********************100%%**********************]  1 of 1 completed


In [9]:
nifty_50_df.head()

Open         High          Low        Close    Adj Close  \
Date                                                                          
2009-04-01  3023.850098  3069.300049  2965.699951  3060.350098  3060.350098   
2009-04-02  3061.050049  3228.750000  3061.050049  3211.050049  3211.050049   
2009-04-06  3211.350098  3303.899902  3211.350098  3256.600098  3256.600098   
2009-04-08  3255.350098  3357.050049  3149.250000  3342.949951  3342.949951   
2009-04-09  3346.000000  3401.149902  3307.050049  3342.050049  3342.050049   

            Volume  
Date                
2009-04-01       0  
2009-04-02       0  
2009-04-06       0  
2009-04-08       0  
2009-04-09       0

Before the change:

|Date|Open|High|Low|Close|Adj Close|Volume|
|---|---|---|---|---|---|---|
|2007-09-17 00:00:00|4518\.4501953125|4549\.0498046875|4482\.85009765625|4494\.64990234375|4494\.64990234375|0|
|2007-09-18 00:00:00|4494\.10009765625|4551\.7998046875|4481\.5498046875|4546\.2001953125|4546\.2001953125|0|
|2007-09-19 00:00:00|4550\.25|4739\.0|4550\.25|4732\.35009765625|4732\.35009765625|0|
|2007-09-20 00:00:00|4734\.85009765625|4760\.85009765625|4721\.14990234375|4747\.5498046875|4747\.5498046875|0|
|2007-09-21 00:00:00|4752\.9501953125|4855\.7001953125|4733\.7001953125|4837\.5498046875|4837\.5498046875|0|

---

After the change:

|Date|Open|High|Low|Close|Adj Close|Volume|
|---|---|---|---|---|---|---|
|2009-04-01 00:00:00|3023\.85009765625|3069\.300048828125|2965\.699951171875|3060\.35009765625|3060\.35009765625|0|
|2009-04-02 00:00:00|3061\.050048828125|3228\.75|3061\.050048828125|3211\.050048828125|3211\.050048828125|0|
|2009-04-06 00:00:00|3211\.35009765625|3303\.89990234375|3211\.35009765625|3256\.60009765625|3256\.60009765625|0|
|2009-04-08 00:00:00|3255\.35009765625|3357\.050048828125|3149\.25|3342\.949951171875|3342\.949951171875|0|
|2009-04-09 00:00:00|3346\.0|3401\.14990234375|3307\.050048828125|3342\.050048828125|3342\.050048828125|0|



In [10]:
nifty_50_df.tail()

Open          High           Low         Close  \
Date                                                                 
2024-03-21  21989.900391  22080.949219  21941.300781  22011.949219   
2024-03-22  21932.199219  22180.699219  21883.300781  22096.750000   
2024-03-26  21947.900391  22073.199219  21947.550781  22004.699219   
2024-03-27  22053.949219  22193.599609  22052.849609  22123.650391   
2024-03-28  22163.599609  22516.000000  22163.599609  22326.900391   

               Adj Close  Volume  
Date                              
2024-03-21  22011.949219  353200  
2024-03-22  22096.750000  388700  
2024-03-26  22004.699219  328400  
2024-03-27  22123.650391  409100  
2024-03-28  22326.900391  410500

## list of stocks in nifty 50

`TO be checked every quarter`

Official site:

* Website for the list: https://www.nseindia.com/products-services/indices-nifty50-index

* CSV download link: https://nsearchives.nseindia.com/content/indices/ind_nifty50list.csv

3rd party site:
Extract from the site:
* https://www.equitymaster.com/india-markets/nse-replica.asp?order=researchText.no_shares_act%20desc

In [44]:
import requests
from bs4 import BeautifulSoup


url = "https://www.equitymaster.com/india-markets/nse-replica.asp?order=researchText.no_shares_act%20desc"
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3"
}

response = requests.get(url, headers=headers)


In [45]:
# Parse the HTML content of the webpage
soup = BeautifulSoup(response.content, "html.parser")

In [46]:
# You need to be able to find this table tag and read all behind it
# Find the div with class "mystocksbig mystocks newtab cls"
table = soup.find('table', {'class': 'mystocksbig mystocks newtab cls'})

In [47]:
type(table)

bs4.element.Tag

In [48]:
# Extract headers
headers = []
for th in table.find('thead').find_all('th'):
    headers.append(th.get_text(strip=True))

In [49]:
headers.remove('Based on Equitymaster Estimatesand Publicly Available Information')
headers[3:]

['COMPANY',
 'INDUSTRY',
 'MARKET PRICE(Rs)',
 'CHANGE(%)',
 'NO OFSHARES(m)',
 'MARKET CAP.**(Rs m)',
 'FREE FLOATADJ. FACTOR#',
 'EARNINGS *(Rs m)',
 'EPS(Rs)',
 'PER(X)']

In [50]:
selected_headers  = ['COMPANY', 'INDUSTRY']

In [51]:
# Extract rows
rows = []
for tr in table.find('tbody').find_all('tr'):
    cells = tr.find_all('td')
    row = [cell.get_text(strip=True) for cell in cells]
    rows.append(row)

In [52]:
# Write to CSV
csv_filename = 'stocks_data.csv'
with open(csv_filename, 'w', newline='', encoding='utf-8') as csvfile:
    csvwriter = csv.writer(csvfile)
    csvwriter.writerow(headers[3:])  # Write headers
    csvwriter.writerows(rows)    # Write data rows

print(f"Data has been written to {csv_filename}")

Data has been written to stocks_data.csv


In [53]:
nifty50__updated_list = pd.read_csv('./stocks_data.csv', sep=',')
nifty50__updated_list

COMPANY         INDUSTRY MARKET PRICE(Rs)  CHANGE(%)  \
0                   IOC           ENERGY            170.8       -0.3   
1                  ONGC           ENERGY            294.2        2.1   
2                   ITC   FOOD & TOBACCO            440.5        1.6   
3            TATA STEEL            STEEL            172.7       -1.1   
4                  NTPC            POWER            377.3       -0.7   
5            POWER GRID            POWER            339.0       -0.1   
6                   SBI          BANKING            858.2       -0.2   
7             HDFC BANK          BANKING          1,630.2       -1.1   
8            ICICI BANK          BANKING          1,231.0       -0.2   
9         RELIANCE IND.           ENERGY          3,213.9        1.2   
10           COAL INDIA           MINING            493.6        0.4   
11        BHARTI AIRTEL          TELECOM          1,433.4        0.3   
12                WIPRO         SOFTWARE            540.9        1.1   
13                 BPCL           ENERGY            300.8       -1.9   
14              INFOSYS         SOFTWARE          1,660.0        0.8   
15                  TCS         SOFTWARE          3,994.0       -0.4   
16          TATA MOTORS      AUTOMOBILES            999.4        0.6   
17            AXIS BANK          BANKING          1,289.3        0.2   
18     HCL TECHNOLOGIES         SOFTWARE          1,523.4        0.3   
19            JSW STEEL            STEEL            937.9       -1.7   
20           SUN PHARMA  PHARMACEUTICALS          1,562.5       -0.4   
21   HINDUSTAN UNILEVER             FMCG          2,585.8        1.5   
22             HINDALCO        ALUMINIUM            694.5       -0.7   
23    ADANI PORTS & SEZ    MISCELLANEOUS          1,474.3       -1.7   
24  HDFC LIFE INSURANCE        INSURANCE            610.2        0.5   
25  KOTAK MAHINDRA BANK          BANKING          1,850.1       -0.1   
26        BAJAJ FINSERV          FINANCE          1,565.6       -0.9   
27                  L&T      ENGINEERING          3,615.7       -0.3   
28                  M&M      AUTOMOBILES          2,868.6       -0.4   
29    ADANI ENTERPRISES    MISCELLANEOUS          3,085.6       -2.0   
30   SBI LIFE INSURANCE        INSURANCE          1,514.0       -1.0   
31        TECH MAHINDRA         SOFTWARE          1,466.6        0.4   
32               NESTLE   FOOD & TOBACCO          2,601.8        1.1   
33         ASIAN PAINTS           PAINTS          2,900.0       -1.2   
34        TATA CONSUMER             FMCG          1,146.9        0.8   
35                TITAN        RETAILING          3,137.0       -4.0   
36                CIPLA  PHARMACEUTICALS          1,490.5       -1.3   
37        INDUSIND BANK          BANKING          1,435.0        0.1   
38               GRASIM         TEXTILES          2,724.1       -0.9   
39        BAJAJ FINANCE          FINANCE          7,068.8       -1.0   
40      SHRIRAM FINANCE          FINANCE          2,802.3       -2.2   
41        MARUTI SUZUKI      AUTOMOBILES         12,039.0       -0.5   
42          LTIMINDTREE         SOFTWARE          5,369.0       -1.0   
43     ULTRATECH CEMENT           CEMENT         11,555.2       -1.2   
44           BAJAJ AUTO      AUTOMOBILES          9,586.9       -0.5   
45        EICHER MOTORS      AUTOMOBILES          4,748.0        0.2   
46   DIVIS LABORATORIES  PHARMACEUTICALS          4,472.2       -3.4   
47        HERO MOTOCORP      AUTOMOBILES          5,532.3       -0.5   
48       DR. REDDYS LAB  PHARMACEUTICALS          6,529.0        0.1   
49     APOLLO HOSPITALS    MISCELLANEOUS          6,313.0       -0.2   

   NO OFSHARES(m) MARKET CAP.**(Rs m)  FREE FLOATADJ. FACTOR#  \
0        14,121.2           2,411,908                     0.5   
1        12,580.3           3,700,489                     0.4   
2        12,489.3           5,501,536                     1.0   
3        12,483.5           2,156,018                     0.7   
4         9,696.7 

# Data Transformations + EDA

# Modelling

# Trading Simulation

# Automation ( from here onwards it should be moved to local)

# Bonus points